Backport upstream PRs for recursive value compilation

lthls opened the pull request:
Some notes:
- I don't know if recursive values are allowed to be local. The compilation scheme doesn't support this, allocating everything on the heap except functions.
- Non-value layouts/sorts should be forbidden for recursive definitions. I don't know if this is the case at the moment. Unlike the previous algorithm, the new one should deal with non-value bindings correctly if they are not actually recursive, but defining an unboxed tuple recursively will not work and should be forbidden. (Example: `let rec r : #{f : _; g : _ } = fun x -> r.g x, fun y -> r.f y`)
- Support for mixed blocks is not there, but should be doable. When this PR or #2380 gets merged, the other should be updated accordingly.

lpw25 commented:
We don't currently support old-to-new pointers on the stack, so local recursive values are not supported. A set of recursive functions can be local however since they are just a single closure block.

lthls commented:
> We don't currently support old-to-new pointers on the stack, so local recursive values are not supported. A set of recursive functions can be stack-allocated however since they are just a single closure block.

Ok. Is this restriction checked in the type-checker, so that I can `fatal_error` in my code, or should I try to write proper error messages ? And is there already a check that all functions in a recursive definition have the same mode ?

lpw25 commented:
Yeah, that’s all handled in type checker

Ekdohibs reviewed (approved):
I carefully reviewed the changes to `from_lambda`, I only gave a quick look at the rest that has already been reviewed upstream.

goldfirere commented:
I'll take a look at this early next week.

goldfirere commented:
I'm in the midst of reviewing this, but timed out. More next week.

mshinwell commented:
cc @ncik-roberts for mixed blocks

ncik-roberts commented:
The assertion is not important. (I commented this earlier, second-guessed myself, looked some more, and am re-posting this now.) It's effectively a sanity check that the caller isn't expecting `caml_alloc_dummy_mixed` to do something it can't reasonably do in bytecode. I wouldn't object to deleting it, though I also wouldn't object to your competing proposal to put in a sentinel value only in bytecode.

lthls commented:
I removed the assertion. I considered duplicating the allocation primitive (`caml_alloc_dummy_mixed` with two arguments and `caml_alloc_dummy_faux_mixed` with only the size), but at least on the compiler side having a single primitive is simpler.

goldfirere commented:
Just to repeat: this is still sitting right near the top of my stack. It's been a very busy few weeks for me, and I unexpectedly had to take most of this week off. I don't think it makes sense for anyone else to jump in at this point to finish, because I have all the context ready to go. Really should happen next week!

lthls commented on ocaml/lambda/value_rec_compiler.ml:
```diff
@@ -327,23 +330,55 @@ let compute_static_size lam =
+
+    (* Note: the primitives below are not present upstream,
+       so they haven't been reviewed *)
+    | Pmakefloatblock (_, _) ->
```
I believe this corresponds to flat float records.
```ocaml
type flat_float_record = { x : float }
(* The following definitions are allowed *)
let rec a = { x = 0. }
let rec b =
  let _ = c in
  { x = 1. }
and c = Some b 
```

goldfirere replied:
Right. My bad about arrays. I still find this odd (my worry is just as valid about records as arrays), but it's not worth spending time getting to the bottom of. If you think this case can be hit, I'm fine with it.

lthls commented on ocaml/lambda/value_rec_compiler.ml:
```diff
@@ -327,23 +330,55 @@ let compute_static_size lam =
+
+    | Pmakeufloatblock (_, _)
+    | Pmake_unboxed_product _ ->
+        dynamic_size () (* Not allowed *)
```
`dynamic_size` calls `fatal_error`. Unless you think this particular case is more likely to happen than the others, I don't think it's worth diverging. The comment means that we could pre-allocate (if we added the right primitives), but since we don't allow it during typing here we can reject it.

goldfirere replied:
Oops. I was wondering in the back of my head what we could do in the dynamic case, but never bothered to look at the definition. Resolving.

lthls commented on ocaml/lambda/value_rec_compiler.ml:
```diff
@@ -0,0 +1,822 @@
+    | Lifused _ -> Constant
+    | Lregion (e, _) ->
+      compute_expression_size env e
+    | Lexclave e ->
```
I'm not familiar enough with the `Lexclave` constructor to know which contexts it is allowed in. If you're confident that such an expression will never appear anywhere in a recursive definition, I can put `dynamic_size` instead (which raises `Misc.Fatal_error` instead of `Assertion_failure`).

goldfirere replied:
"never appear anywhere": It could be in a function, like `let rec f x = exclave_ ...`. But there can't be a `let rec x = exclave_ ...`.

lthls commented on ocaml/lambda/value_rec_compiler.ml:
```diff
@@ -0,0 +1,837 @@
+  Lambda.lambda split_result =
+  match lam with
+  | Lvar _v ->
+    (* CR vlaviron: We cannot eta-expand when arbitrary layouts are involved. *)
```
The note in comments is the key. Upstream we can generate an eta-expansion without knowing the arity, because everything is a value, but we could generate better code if we propagated the arity.
Here we can't generate the eta-expansion without knowing the arity (including layouts), so I had to implement the arity propagation (it's in commit f2a8efca23158caf216d25df15beb0f574b891c5).
I guess we could push this upstream too, since it's done anyway.

goldfirere replied:
I was reviewing commit-by-commit and forgot to delete this comment before posting. It's all fine as is.

lthls commented on ocaml/lambda/translmod.ml:
```diff
@@ -268,7 +268,7 @@ let record_primitive = function
    upstreamed.  At that point this helper can move into that module. *)
 
-let preallocate_letrec ~bindings ~body =
+let preallocate_class_letrec ~bindings ~body =
```
I don't think so. There isn't much to review, if you haven't done it already I can find someone to look at this.

goldfirere replied:
No, I didn't look at this.

lthls replied:
Resolving (reviewed by @Ekdohibs)

Ekdohibs reviewed (approved):
I especially reviewed `transmod.ml`, given the rest was reviewed upstream.

lthls commented:
@goldfirere I've changed the `Lexclave` case as you suggested in the last commit. I've added a comment to explain why it's different from the other cases.
If you're happy with that I think we can merge.
